File: blurt_py_80211/qam.py
#!/usr/bin/env python
import numpy as np
import util

def grayRevToBinary(x, n):
    y = 0*x
    for i in range(n):
        y <<= 1
        y |= x&1
        x >>= 1
    shift = 1
    while shift<n:
        y ^= y >> shift
        shift<<=1
    return y

class Constellation:
    def __init__(self, Nbpsc, symbols):
        self.Nbpsc = Nbpsc
        self.symbols = symbols

def qam_constellation(Nbpsc):
    scale = 1.
    n = 1
    if Nbpsc != 1:
        # want variance = .5 per channel
        # in fact, var{ 2 range(2^{Nbpsc/2}) - const } = (2^Nbpsc - 1) / 3
        scale = ( 1.5 / ((1<<Nbpsc) - 1) ) ** .5
        n = Nbpsc >> 1
    symbols = (2*grayRevToBinary(np.arange(1<<n), n) + 1 - (1<<n)) * scale
    if Nbpsc != 1:
        symbols = np.tile(symbols, 1<<n) + 1j*np.repeat(symbols, 1<<n)
    return Constellation(Nbpsc, symbols)

bpsk = qam_constellation(1)
qpsk = qam_constellation(2)
qam16 = qam_constellation(4)
qam64 = qam_constellation(6)

def encode(interleaved_bits, rate):
    return rate.constellation[util.shiftin(interleaved_bits, rate.Nbpsc)]

def demapper(data, constellation, dispersion, n):
    squared_distance = np.abs(constellation[np.newaxis,:] - data[:,np.newaxis])**2
    ll = -np.log(np.pi * dispersion) - squared_distance / dispersion
    ll -= np.logaddexp.reduce(ll, 1)[:,np.newaxis]
    j = np.arange(1<<n)
    ll0 = np.zeros((data.shape[0], n), float)
    ll1 = np.zeros((data.shape[0], n), float)
    for i in xrange(n):
        idx0 = np.where(0 == (j & (1<<i)))[0]
        idx1 = np.where(j & (1<<i))[0]
        ll0[:,i] = np.logaddexp.reduce(ll[:,idx0], 1)
        ll1[:,i] = np.logaddexp.reduce(ll[:,idx1], 1)
    return np.int64(np.clip(10.*(ll1-ll0), -1e4, 1e4)).flatten()

File: blurt_cpp_80211/qam.cc
#include "qam.h"
#include <cmath>
#include <stdint.h>

uint32_t grayRevToBinary(uint32_t x, uint32_t n) {
    uint32_t y = 0;
    for (size_t i=0; i<n; i++) {
        y <<= 1;
        y |= x&1;
        x >>= 1;
    }
    for (size_t shift=1; shift<n; shift<<=1)
        y ^= y >> shift;
    return y;
}

QAM::QAM(uint32_t Nbpsc) : Constellation(Nbpsc) {
    float scale = 1.f;
    uint32_t n = 1;
    if (Nbpsc != 1) {
        // want variance = .5 per channel
        // in fact, var{ 2 range(2^{Nbpsc/2}) - const } = (2^Nbpsc - 1) / 3
        scale = sqrtf( 1.5f / ((1<<Nbpsc) - 1) );
        n = Nbpsc >> 1;
    }
    std::vector<float> syms(1<<n);
    for (uint32_t i=0; i<1<<n; i++)
        syms[i] = (2*int(grayRevToBinary(i, n)) + 1 - (1<<n)) * scale;
    if (Nbpsc != 1) {
        size_t k=0;
        for (size_t i=0; i<1<<n; i++)
            for (size_t j=0; j<1<<n; j++)
                symbols[k++] = complex(syms[j], syms[i]);
    } else {
        for (size_t i=0; i<1<<n; i++)
            symbols[i] = syms[i];
    }
}
